core/io.go

```diff
@@ -0,0 +1,98 @@
+package core
+
+import (
+	"bytes"
+	"errors"
+	"io"
+
+	"github.com/dicedb/dice/config"
+)
+
+type RESPParser struct {
+	c   io.ReadWriter
+	buf *bytes.Buffer
+}
+
+func NewRESPParser(c io.ReadWriter) *RESPParser {
+	return NewRESPParserWithBytes(c, []byte{})
+}
+
+func NewRESPParserWithBytes(c io.ReadWriter, initBytes []byte) *RESPParser {
+	var b []byte
+	var buf *bytes.Buffer = bytes.NewBuffer(b)
+	buf.Write(initBytes)
+	return &RESPParser{
+		c:   c,
+		buf: buf,
+	}
+}
+
+func (rp *RESPParser) DecodeOne() (interface{}, error) {
+	// assigning temporary buffer to read 512 bytes in one shot
+	// and reading them in a loop until we have all the data
+	// we want.
+	// note: the size 512 is arbitrarily chosen, and we can put
+	// a decent thought into deciding the optimal value (in case it affects the perf)
+	var buf []byte = make([]byte, config.IOBufferLength)
+
+	// keep reading the bytes from the buffer until the first /r is found
+	// note: there may be extra bytes read over the socket post /r
+	// but that is fine.
+	for {
+		n, err := rp.c.Read(buf)
+		if n <= 0 {
+			break
+		}
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		rp.buf.Write(buf[:n])
+
+		if bytes.Contains(buf, []byte{'\r', '\n'}) {
+			break
+		}
+	}
+
+	b, err := rp.buf.ReadByte()
+	if err != nil {
+		return nil, err
+	}
+
+	switch b {
+	case '+':
+		return readSimpleString(rp.c, rp.buf)
+	case '-':
+		return readError(rp.c, rp.buf)
+	case ':':
+		return readInt64(rp.c, rp.buf)
+	case '$':
+		return readBulkString(rp.c, rp.buf)
+	case '*':
+		return readArray(rp.c, rp.buf, rp)
+	}
+
+	// this also captures the Cross Protocol Scripting attack.
+	// Since we do not support simple strings, anything that does
+	// not start with any of the above special chars will be a potential
+	// attack.
+	// Details: https://bou.ke/blog/hacking-developers/
+	return nil, errors.New("possible cross protocol scripting attack detected")
```

Comment:
Already discussed on Discord -> https://discordapp.com/channels/1034342738960855120/1034346443927990294/1034483641100795954

That needs a lot of refactoring, for now we can concentrate on features, later we can start error codes and error handler.

Reply:
@rohanverma94 yup, I was aware of the discussion happened. When I saw the error handling here, it reminded me of this style-guide I had read once, so added it. Let's create an issue and add all our findings so that when we get to pick it up, we'll have some decent references to get started.